qpsolvers/solvers/hpipm_.py

```diff
@@ -0,0 +1,273 @@
+arising in model predictive control and embedded optmization. If you are using
+HPIPM in some academic work, consider citing the corresponding paper
+[Frison2020]_.
+"""
+
+import warnings
+from typing import Optional
+
+import numpy as np
+import hpipm_python.common as hpipm
+
+from ..problem import Problem
+from ..solution import Solution
+
+
+def hpipm_solve_problem(
+    problem: Problem,
+    initvals: Optional[np.ndarray] = None,
+    mode: str = "balance",
+    verbose: bool = False,
+    **kwargs,
+) -> Solution:
+    """Solve a quadratic program using HPIPM.
+
+    Parameters
+    ----------
+    problem :
+        Quadratic program to solve.
+    initvals :
+        Warm-start guess vector.
+    mode :
+        Solver mode, which provides a set of default solver arguments. Pick one
+        of ["speed_abs", "speed", "balance", "robust"]. Default is "balance".
+    verbose :
+        Set to `True` to print out extra information.
+
+    Returns
+    -------
+    :
+        Solution returned by the solver.
+
+    Notes
+    -----
+    Keyword arguments are forwarded to HPIPM. For instance, we can call
+    ``hpipm_solve_qp(P, q, G, h, u, tol_eq=1e-5)``. HPIPM settings include the
+    following:
+
+    .. list-table::
+       :widths: 30 70
+       :header-rows: 1
+
+       * - Name
+         - Description
+       * - ``iter_max``
+         - Maximum number of iterations.
+       * - ``tol_eq``
+         - Equality constraint tolerance.
+       * - ``tol_ineq``
+         - Inequality constraint tolerance.
+       * - ``tol_comp``
+         - Complementarity condition tolerance.
+       * - ``tol_stat``
+         - Stationarity condition tolerance.
+    """
+    P, q, G, h, A, b, lb, ub = problem.unpack()
+
+    # setup the problem dimensions
+    nv = q.shape[0]
+    ne = b.shape[0] if b is not None else 0
+    ng = h.shape[0] if h is not None else 0
+
+    nlb = lb.shape[0] if lb is not None else 0
+    nub = ub.shape[0] if ub is not None else 0
+    nb = max(nlb, nub)
+
+    dim = hpipm.hpipm_dense_qp_dim()
+    dim.set("nv", nv)
+    dim.set("nb", nb)
+    dim.set("ne", ne)
+    dim.set("ng", ng)
+
+    # setup the problem data
+    qp = hpipm.hpipm_dense_qp(dim)
+    qp.set("H", P)
+    qp.set("g", q)
+
+    if ng > 0:
+        qp.set("C", G)
+        qp.set("ug", h)
+        # mask out the lower bound
+        qp.set("lg_mask", np.zeros_like(h))
+
+    if ne > 0:
+        qp.set("A", A)
+        qp.set("b", b)
+
+    if nb > 0:
+        # mark all variables as box-constrained
+        qp.set("idxb", np.arange(nv))
+
+        # need to mask out lb or ub if the box constraints are only one-sided
+        # we also mask out infinities (and set the now-irrelevant value to
+        # zero), since HPIPM doesn't like infinities
+        if nlb > 0:
+            lb_mask = np.isinf(lb)
+            lb[lb_mask] = 0
+            qp.set("lb", lb)
+            qp.set("lb_mask", ~lb_mask)
+        else:
+            qp.set("lb_mask", np.zeros(nb))
```

Comment:
explicit typing, to highlight this is a boolean array:
```suggestion
            qp.set("lb_mask", np.zeros(nb, dtype=bool))
```